20 & 21 - skip metadata pipeline

poikilotherm opened the pull request:
Closes #20 
Closes #21

TODOs
- [x] Add tests for Metadata methods
- [x] Add tests for InMemoryItemRepo & GetRecordHandler & ListRecordsHandler with CopyElement backing

poikilotherm committed:
refactor(xmlio): expose underlying XmlStreamReader API hasNext() and nextEvent()

This is necessary to use these basic routines within the EchoElement,
as we will not read from a String there, but make it capable to read from
an InputStream, too.

poikilotherm committed:
refactor(common): make EchoElement capable of writing from InputStream

Before, the XML string sent to the EchoElement was stuffed into a XmlReader
with a ByteArrayInputStream. Now we extend this to be capable of reading
from an arbitrary InputStream when given on object creation.

This commit also adds benchmarking tests using JMH to learn about the speed
decrease that parsing the XML from the input stream causes. It get's compared
to the "native" copy of input to output as seen in Dataverse. Run it via
`mvn -Pbenchmark clean test`

Another small change happened, too: instead of the deprecated Stack class
we now use the replacement "Deque".

poikilotherm committed:
doc(common): add Javadocs to EchoElementBenchmark

poikilotherm committed:
style(xmlio): remove unused import from XmlIoWriter

poikilotherm committed:
feat(common): add ReplacingInputStream to filter XML declaration

This implementation of FilterInputStream has been copied from Apache POI
or, more precise, its origin at Inbot.

Currently, the Dataverse OAI-PMH data provider uses this filter
to remove the XML declaration from the pregenerated XML metadata files.

It is being added here, but flagged as deprecated, purely for
benchmarking reasons.

poikilotherm committed:
feat(common): add CopyElement to paste XML stream data

This is a fast alternative to EchoElement, which does not do
any XML parsing before it copies XML data from an InputStream
into the XmlWriter.

It takes care of tricking the writer into accepting the data
without further addo, but REQUIRES that the writer already
contains a wrapping element (you cannot write at root with this).

poikilotherm committed:
test(common): extend benchmarking for CopyElement and compare with present Dataverse usage

poikilotherm committed:
docs(common): add deprecation Javadoc to ReplacingInputStream

poikilotherm committed:
refactor(common): change Metadata to use EchoElement, CopyElement or XOAIMetadata

All of these share the common interface XmlWritable, so we can store the element to write out
as metadata as this type (no switch or if necessary).

This also moves the data handling to using classes - they create the data and the XML writables,
this class is just used for the modeling of a fluent API.

It is linked to the creation of items, which are generated by the application using this library
via the repository interfaces. The application may decide how to fill in metadata when creating items.

poikilotherm committed:
refactor(common): make XSLPipeline.apply() return gracefully

When adding new transformers, ignore null ones and proceed.
Simplifies adding transformers from context and metadata format in data-provider.

poikilotherm committed:
refactor(data): adapt Item interface to return empty about list by default

Instead of having to override the function unnecessary, provide a default of return an empty list

poikilotherm committed:
refactor(data): cleanup common exceptions in ItemRepositoryHelper

poikilotherm committed:
refactor(data): specify return types in ItemRepository and helper #21

Instead of an Item, make the methods sending an identifier only, return an object of type
ItemIdentifier to make it more clear this does not carry metadata - this interface does not expose
a getMetadata(), which is done with the Item interface only.

The getItem() method is changed to send along a metadata format, so the application
can expose pregenerated or cached metadata within the specific format.

See also classes CopyElement and Metadata.copyFromStream()

poikilotherm committed:
refactor(data): make record handlers skip processing #20

- Shell out the XSL pipeline handling to a new MetadataHelper to avoid
  duplicated code in GetRecordHandler and ListRecordsHandler
- Send metadata format via ItemRepository.getItem(identifier, format)
  to retrieve an item filled with metadata
- The refactored Metadata class allows to distinguish if the underlying
  data needs processing or not. Reusing this here to skip the XSL
  pipeline when unnecessary. It's up to the application to provide
  valid metadata that create a validatable OAI-PMH response!
- Skipping the processing allows for pregeneration/caching of
  potentially large metadata like DDI codebooks

poikilotherm commented:
@landreev  can you please review?

poikilotherm commented:
@landreev  can you please review?

poikilotherm committed:
style(data) remove some unused imports

poikilotherm committed:
test: refactor testing for new Metadata handling

- Refactor InMemoryItem and InMemoryItemRepo with a more consice API
- Extend the GetRecordHandler and ListRecordsHandler tests:
    - Add explicit test example cases that include non-deleted, random metadata items
    - Add explicit test example cases that include non-deleted items
      associated with a CopyElement and InputStream
    - Verify the correct existance, but do not validate the OAI-PMH response
      (yet?)

poikilotherm committed:
test(common): check for null transformer in XSLPipeline.apply()

poikilotherm committed:
ci: make Sonarcloud use aggregated JaCoCo coverage report

landreev commented on xoai-common/src/main/java/io/gdcc/xoai/util/ReplacingInputStream.java:
```diff
@@ -0,0 +1,195 @@
+ * Slightly adapted to avoid Apache Commons Lang3 dependency
+ *
+ * @author jillesvangurp
+ * @deprecated This should not be used anywhere - it's very slow and of limited use for XOAI
```
Would it make sense to add another sentence here, spelling out that it's here for benchmarking/comparison only? - I was momentarily confused seeing it in the PR.

poikilotherm replied:
Yes, absolutely.

I will also move it from `main` to `test`, so no one uses it accidentally.

poikilotherm replied:
Done with f131d30

landreev commented on xoai-common/src/main/java/io/gdcc/xoai/xml/CopyElement.java:
```diff
@@ -0,0 +1,98 @@
+     * we do not interfere with any XML processing instructions following.
+     */
+    private static final Matcher xmlDeclaration = Pattern.compile("<\\?xml .*?\\?>").matcher("");
+    
```
is `<\\?xml[^>]*\\?>`  any better/prettier?

poikilotherm replied:
Let's use a saying here: "Geschmackssache", sagte der Affe und bis in die Schmierseife. ("Matter of taste", monkey said and took a bite from the soap.)

Not sure which is faster, both work.

landreev reviewed (approved):
Super impressive work! 
Merge it merge it merge it :smile:

poikilotherm committed:
style(common): move ReplacingInputStream to test package

No one should use this FilterInputStream, so we exclude it from
being shipped with the xoai-common JAR. As it's only being used
for EchoElementBenchmark, it can happily live within src/test
